net/tcp packet dropping reliability

ssharks opened the pull request:
During testing with artificial packet loss a few problems surfaced namely:

- A deadlock in updating the receive window size
- Silently failing the resend due to insufficient buffer size
- The send timer being canceled by new data being transmitted for which no buffer was available
- Incorrect computation of the transmit MSS

This PR fixes the issues above

ssharks committed:
net: tcp: Use our MTU to determine the mss for transmission

In the existing the value received from the other side by the TCP options
is used as MSS for transmission. Since the MSS options are an
announcement rather then a negotionation, it is likely the receiver will
have a different and possibly bigger MSS than allowed by our side.
This allow potentially for different a MSS in the receive and transmit
path.

Directly using the received MSS could cause problems when our MSS is only
allowed to be small. At transmission, for that reason take the minimum of
the received MSS and our desired MSS to find a value compatible to both
sides of the link.

Rename the function to net_tcp_get_recv_mss to net_tcp_get_supported_mss
to better reflect its function in the new situation.

Signed-off-by: Sjors Hettinga <s.a.hettinga@gmail.com>

ssharks committed:
net: tcp: Avoid deadlock in updating the TCP receive window

The function tcp_data_get tries to update the TCP receive window using
net_context_update_recv_wnd. This function graps the context lock while
the tcp_data_get is called from a situation where it already has tcp lock
is already. Transmission actions do first grab the the context lock and
try to grab the tcp lock afterwards. The combination of both can cause a
deadlock.

By taking the shortcut to directly update the tcp receive window without
going through the net context, the context lock is not required avoiding a
possible deadlock situation.

Signed-off-by: Sjors Hettinga <s.a.hettinga@gmail.com>

ssharks committed:
net: tcp: Avoid new packets from causing killing the pending send_timer

When there was no room to transmit the a next packet to be transmitted,
the -ENOBUFS could cause the retransmission to fail.

Secondly the conn->unacked_len can be set to 0 in the retransmission
process, causing the subscribtion to the transmit timer to fail. Use the
variable send_data_total instead.

Make sure that is the send_data buffer becomes empty the send_timer is
cancelled, but make sure any pending data still keeps on being transmitted.

Signed-off-by: Sjors Hettinga <s.a.hettinga@gmail.com>

ssharks committed:
net: tcp: Explicitly notify buffer allocation failures in retransmission

Log an error to explicitly log a failed buffer allocation in TCP
retransmission. This avoids silently failing retransmissions due to
repeating buffer allocation failures.

Signed-off-by: Sjors Hettinga <s.a.hettinga@gmail.com>

ssharks commented:
@AndreyDodonov-EH You might be interested in this one as well

ssharks committed:
net: tcp: Fix the polling implementation with corrected window filling

The window full computation was corrected to use the send_total instead
of the unacked_len. This conflicted with the new polling implementation
due to the moment when these values are changed.

Move taking the tx_sem outside of tcp_send_queued_data to handle the
-ENOBUF situation properly in case called from net_tcp_queue_data.
net_tcp_queue_data removes data from the send_data in case the
transmission failed with -ENOBUF. This cause the buffer to be not full
any more.

Signed-off-by: Sjors Hettinga <s.a.hettinga@gmail.com>

rlubos commented on subsys/net/ip/tcp.c:
```diff
@@ -1092,19 +1097,23 @@ static int tcp_send_queued_data(struct tcp *conn, size_t new_packet_size)
 
 	if (tcp_window_full(conn)) {
-		(void)k_sem_take(&conn->tx_sem, K_NO_WAIT);
+		/* When new data is transmitted it will be removed from the list
+		 * afterwards, do not take the semaphore
```
Then I think this function should be moved out of this function, and window should be verified from within `net_tcp_queue_data()`/`tcp_in()` separately.

ssharks replied:
Makes a lot of sense, this simplifies the is logic a ton, will fix it

rlubos commented on subsys/net/ip/tcp.c:
```diff
@@ -1117,10 +1117,15 @@ static int tcp_send_queued_data(struct tcp *conn)
 
 	/* If we have out-of-bufs case, then do not start retransmit timer
 	 * yet. The socket layer will catch this and resend data if needed.
+	 * Only perform this when it is just the newly added packet,
```
It's not clear to me why is ths a problem when `tcp_send_queued_data()` is called from within `tcp_in()` - this timer should've been canceled already:
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/ip/tcp.c#L2033

And I don't see where could it be rescheduled again inbetween.

ssharks replied:
Hmmm.... maybe you are right

ssharks replied:
Thinking about it a little longer, there is still an option when data is the queue and a new packet is added from net_tcp_queue_data. In that case the failing new packet could potentially kill the retransmit timer, where it should be still running.

ssharks replied:
I can simplify this commitment by moving the cancel statement into net_tcp_queue_data. The send timer can be cancelled when the send_data_total reaches 0 after an -ENOBUFS error.

ssharks replied:
The simplification has been implemented in https://github.com/zephyrproject-rtos/zephyr/pull/45890/commits/c6b1973d5567b7d3f7f96793ac5fb54c0c3213ab. It required some magic rebasing to get the other 2 commits also properly in. The end result is a much cleaner tcp_send_queued_data function.

@rlubos Currently the send_timer can be triggered if there is no data in the transmission queue and the transmission window is 0. Do you think that is correct? Otherwise we have to remove 
'if (tcp_window_full(conn)) {
		subscribe = true;
}'
from tcp_send_queued_data

rlubos replied:
I don't think there's a point scheduling the retransmission timer with zero-length window, I faced similar issue here:
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/ip/tcp.c#L2285
IMO in such case we need to either wait for an ACK from the peer side (w/o scheduling retransmission), or rely on the ZWP timer and send ZWP when neeeded.

rlubos commented on subsys/net/ip/tcp.c:
```diff
@@ -997,6 +1020,11 @@ static int tcp_unsent_len(struct tcp *conn)
 	}
 
 	unsent_len = conn->send_data_total - conn->unacked_len;
+	if (conn->unacked_len >= conn->send_win) {
+		unsent_len = 0;
```
When could this happen? Peer reduced RX window size on his side?

ssharks replied:
Indeed, although unlikely it could happen.

rlubos commented on subsys/net/ip/tcp.c:
```diff
@@ -997,6 +1020,11 @@ static int tcp_unsent_len(struct tcp *conn)
 	}
 
 	unsent_len = conn->send_data_total - conn->unacked_len;
+	if (conn->unacked_len >= conn->send_win) {
```
Generally, the concern I have regarding this change, is that we modify the `tcp_unsent_len()` function content, so that it does not always return true - as it now takes the remaining window size into account as well. So we don't really return the amount of data that we should send, but rather the amount we can. I find this a little confusing, but it might be just me.

Given that, I wonder if it wouldn't be a better approach to leave this function intact, and rather modify `send_data_total ` value only after successful send - that way `tcp_window_full()` could be used to retrieve the window status regardless of the context it is called in.

ssharks replied:
> I just looked at the function `net_tcp_queue_data`. When an -ENOBUFS by `tcp_send_queued_data` is returned, it throws away the whole block of size `len` from the `send_data`. If the `len` is > MSS, it could happen that the first section is transmitted, but at the second an -ENOBUFS occurs. In that case the data is transmitted, but later on removed from the send_data. What this will do to the sequence number administration is a big question.

Just added a possible fix for this potential issue.

ssharks commented:
Rebased on top of main

ssharks committed:
net: tcp: Fix accidental removal of data from send_data

In the function net_tcp_queue_data. When an -ENOBUFS by
tcp_send_queued_data is returned, it throws away the whole block of size
len from the send_data. If the len is > MSS, it could happen that the
first section is transmitted, but at the second an -ENOBUFS occurs.
In that case the data is transmitted, but later on removed from the
send_data.

To circumvent this problem, check if the len + unacked_len is smaller
then the send_data_total. If so, the data can safely be removed from
send_data. Otherwise, just pretend the transmission went OK. The
acknowledgment and retransmit path will eventually take care of it.

Signed-off-by: Sjors Hettinga <s.a.hettinga@gmail.com>

ssharks commented on subsys/net/ip/tcp.c:
```diff
@@ -2351,6 +2351,19 @@ int net_tcp_queue_data(struct net_context *context, struct net_pkt *pkt)
 
+	if ((ret == -ENOBUFS) &&
+		(conn->send_data_total < (conn->unacked_len + len))) {
+		/* Some of the data has been send, we cannot remove the
```
Fixed

ssharks commented:
@jukkar Could you have a look at this PR?

rlubos reviewed (approved):
No more issues from my side.

ssharks commented:
> No more issues from my side.

Thanks for looking into it
